diff --git a/flagscale/train/train_qwen2_5_vl.py b/flagscale/train/train_qwen2_5_vl.py
index 80b8521c..73958bd3 100644
--- a/flagscale/train/train_qwen2_5_vl.py
+++ b/flagscale/train/train_qwen2_5_vl.py
@@ -14,77 +14,36 @@
 # limitations under the License.
 
 
-import logging
 import os
 import sys
-
-from argparse import Namespace
-from copy import deepcopy
 from functools import partial
-from typing import List, Optional, Tuple, Union
+from copy import deepcopy
+from typing import Union, Optional, Tuple
 
 import torch
 import torch._dynamo
 
-from megatron.core import parallel_state
-from megatron.core.datasets.blended_megatron_dataset_builder import BlendedMegatronDatasetBuilder
-from megatron.core.datasets.gpt_dataset import GPTDataset, GPTDatasetConfig, MockGPTDataset
-from megatron.core.enums import ModelType
-from megatron.core.models.gpt import GPTModel
-from megatron.core.models.gpt.gpt_layer_specs import (
-    get_gpt_decoder_block_spec,
-    get_gpt_layer_local_spec,
-    get_gpt_layer_with_transformer_engine_spec,
-    get_gpt_mtp_block_spec,
-)
-from megatron.core.models.gpt.heterogeneous.heterogeneous_layer_specs import (
-    get_gpt_heterogeneous_layer_spec,
-)
-from megatron.core.rerun_state_machine import get_rerun_state_machine
-from megatron.core.transformer.spec_utils import import_module
-from megatron.core.utils import StragglerDetector
-from megatron.training import get_args, get_timers, get_tokenizer, print_rank_0
-from megatron.training.arguments import core_transformer_config_from_args
-from megatron.training.checkpointing import get_checkpoint_name  # for dataloder
-from megatron.training.utils import (
-    get_batch_on_this_cp_rank,
-    get_batch_on_this_tp_rank,
-    get_blend_and_blend_per_split,
-)
-from megatron.training.yaml_arguments import core_transformer_config_from_yaml
+from argparse import Namespace
 
 # # For pytorch 2.6
 # torch.serialization.add_safe_globals([Namespace])
 
+from megatron.core import mpu
 
-import megatron.legacy.model  # isort: skip
-
-# NOTE: Loading `megatron.legacy.model` earlier fails due to circular import
-
-try:
-    from megatron.post_training.arguments import add_modelopt_args, modelopt_args_enabled
-    from megatron.post_training.loss_func import loss_func as loss_func_modelopt
-    from megatron.post_training.model_provider import model_provider as model_provider_modelopt
-
-    has_nvidia_modelopt = True
-except ImportError:
-    has_nvidia_modelopt = False
-
-from flagscale.train.datasets.sft_dataset import SFTDataset, SFTDatasetConfig
-from flagscale.train.extra_valid import extra_valid_datasets_provider
-from flagscale.train.train import pretrain
-
-stimer = StragglerDetector()
-
-#### especially for qwen2.5-vl ####
+from megatron.core import parallel_state
+from megatron.training.checkpointing import get_checkpoint_name
+from megatron.core.enums import ModelType
+from megatron.training import get_args, get_timers, pretrain, print_rank_0
+from megatron.training.arguments import core_transformer_config_from_args
+from megatron.training.utils import (
+    average_losses_across_data_parallel_group,
+)
 from megatron.core.num_microbatches_calculator import get_num_microbatches
 
+
 torch._dynamo.config.suppress_errors = True
-from megatron.core.parallel_state import (
-    get_pipeline_model_parallel_rank,
-    get_pipeline_model_parallel_world_size,
-    get_tensor_model_parallel_rank,
-)
+from megatron.core import mpu
+from megatron.core.parallel_state import get_tensor_model_parallel_rank, get_pipeline_model_parallel_world_size, get_pipeline_model_parallel_rank
 from megatron.energon import (
     LimitDataset,
     RepeatDataset,
@@ -94,27 +53,23 @@ from megatron.energon import (
     get_train_dataset,
     get_val_datasets,
 )
-from megatron.training.global_vars import get_tokenizer
+
 from megatron.training.tokenizer.tokenizer import build_tokenizer
-from tools.datasets.qwenvl.data.dataset_helpers import TaskEncoder, print_error_handler
+from megatron.training.global_vars import get_tokenizer
 
 from flagscale.train.models.qwen2_5_vl.layer_specs import (
     get_gpt_layer_with_transformer_engine_spec,
-    get_mlp_module_spec,
     get_qwen2vl_vision_model_spec,
+    get_mlp_module_spec
+
 )
 from flagscale.train.models.qwen2_5_vl.qwen2_5_vl_model import Qwen2_5VLModel
 from flagscale.train.models.qwen2_5_vl.tensor_parallel import broadcast_data
 from flagscale.train.models.qwen2_5_vl.transformer_config import (
     get_vision_model_config,
-    get_vision_projection_config,
+    get_vision_projection_config
 )
-
-#### especially for qwen2.5-vl ####
-IGNORE_IDX = -100
-FIRST_MAX_PADDING_FLAG = True
-LAST_LARGE_IMG = False
-
+from tools.datasets.qwenvl.data.dataset_helpers import TaskEncoder, print_error_handler
 
 def model_provider(
     pre_process=True, post_process=True, add_encoder=True, add_decoder=True
@@ -129,47 +84,45 @@ def model_provider(
     if not use_te:
         raise NotImplementedError("The Qwen2-VL model is only implemented with TransformerEngine!")
 
-    if (
-        args.rotary_seq_len_interpolation_factor is not None
-        or args.rotary_seq_len_interpolation_factor != 1
-    ):
+    if args.rotary_seq_len_interpolation_factor is not None or args.rotary_seq_len_interpolation_factor != 1:
         print_rank_0('Multimodal RoPE currently not support RoPE interpolation, set to None...')
         args.rotary_seq_len_interpolation_factor = None
 
     vision_config = get_vision_model_config(args, deepcopy(config))
     vision_config.pipeline_model_parallel_size = 1
     vision_config.first_pipeline_num_layers = None
-    vision_projector_config = get_vision_projection_config(
-        deepcopy(config), vision_config.hidden_size, args.spatial_merge_size
-    )
+    vision_projector_config = get_vision_projection_config(deepcopy(config), vision_config.hidden_size, args.spatial_merge_size)
 
     print_rank_0("building Qwen2-5-VL model in TE...")
     # Layer Specs of vit, llm and projector
     transformer_layer_spec = get_gpt_layer_with_transformer_engine_spec(args.qk_layernorm)
     vision_model_spec = get_qwen2vl_vision_model_spec()
     vision_projector_spec = get_mlp_module_spec(add_norm=False).submodules
-    if args.enable_variable_seq_lengths:
-        config.variable_seq_lengths = True
 
     model = Qwen2_5VLModel(
         language_transformer_config=config,
         language_transformer_layer_spec=transformer_layer_spec,
         language_vocab_size=args.padded_vocab_size,
         language_max_sequence_length=args.max_position_embeddings,
+
         vision_transformer_config=vision_config,
         vision_transformer_layer_spec=vision_model_spec,
-        drop_vision_class_token=False,  # NOTE: no class token to drop?
+        drop_vision_class_token=False, # NOTE: no class token to drop?
+
         vision_projection_config=vision_projector_config,
         vision_projection_layer_spec=vision_projector_spec,
         vision_projection_type='mlp',
-        allow_missing_vision_projection_checkpoint=args.allow_missing_vision_projection_checkpoint,
+        allow_missing_vision_projection_checkpoint= args.allow_missing_vision_projection_checkpoint,
+
         language_position_embedding_type=args.position_embedding_type,
         language_rotary_percent=args.rotary_percent,
         language_rotary_base=args.rotary_base,
+
         pre_process=pre_process,
         post_process=post_process,
         add_decoder=add_decoder,
         add_encoder=add_encoder,
+
         fp16_lm_cross_entropy=args.fp16_lm_cross_entropy,
         parallel_output=True,
         language_share_embeddings_and_output_weights=not args.untie_embeddings_and_output_weights,
@@ -178,13 +131,12 @@ def model_provider(
     model.freeze(
         freeze_language_model=args.freeze_LM,
         freeze_vision_model=args.freeze_ViT,
-        freeze_vision_projection=False,
+        freeze_vision_projection=False
     )
 
     return model
 
-
-# copy from https://github.com/huggingface/transformers/blob/40a493c7ed4f19f08eadb0639cf26d49bfa5e180/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L1404
+# Slightly modified from Qwen2_5VLForConditionalGeneration.get_rope_index
 def get_rope_index(
     input_ids: Optional[torch.LongTensor] = None,
     image_grid_thw: Optional[torch.LongTensor] = None,
@@ -331,21 +283,9 @@ def get_rope_index(
                 time_tensor_long = time_tensor.long()
                 t_index = time_tensor_long.flatten()
 
-                h_index = (
-                    torch.arange(llm_grid_h)
-                    .view(1, -1, 1)
-                    .expand(llm_grid_t, -1, llm_grid_w)
-                    .flatten()
-                )
-                w_index = (
-                    torch.arange(llm_grid_w)
-                    .view(1, 1, -1)
-                    .expand(llm_grid_t, llm_grid_h, -1)
-                    .flatten()
-                )
-                llm_pos_ids_list.append(
-                    torch.stack([t_index, h_index, w_index]) + text_len + st_idx
-                )
+                h_index = torch.arange(llm_grid_h).view(1, -1, 1).expand(llm_grid_t, -1, llm_grid_w).flatten()
+                w_index = torch.arange(llm_grid_w).view(1, 1, -1).expand(llm_grid_t, llm_grid_h, -1).flatten()
+                llm_pos_ids_list.append(torch.stack([t_index, h_index, w_index]) + text_len + st_idx)
                 st = ed + llm_grid_t * llm_grid_h * llm_grid_w
 
             if st < len(input_tokens):
@@ -356,9 +296,7 @@ def get_rope_index(
             llm_positions = torch.cat(llm_pos_ids_list, dim=1).reshape(3, -1)
             position_ids[..., i, attention_mask[i] == 1] = llm_positions.to(position_ids.device)
             mrope_position_deltas.append(llm_positions.max() + 1 - len(total_input_ids[i]))
-        mrope_position_deltas = torch.tensor(
-            mrope_position_deltas, device=input_ids.device
-        ).unsqueeze(1)
+        mrope_position_deltas = torch.tensor(mrope_position_deltas, device=input_ids.device).unsqueeze(1)
         return position_ids, mrope_position_deltas
     else:
         if attention_mask is not None:
@@ -374,21 +312,22 @@ def get_rope_index(
                 .expand(3, input_ids.shape[0], -1)
             )
             mrope_position_deltas = torch.zeros(
-                [input_ids.shape[0], 1], device=input_ids.device, dtype=input_ids.dtype
+                [input_ids.shape[0], 1],
+                device=input_ids.device,
+                dtype=input_ids.dtype,
             )
 
         return position_ids, mrope_position_deltas
 
-
 def get_ltor_masks_and_position_ids(
-    input_ids,
-    image_thw_grids,
-    video_thw_grids,
-    target,
-    pad_token,
-    second_per_grid_ts,
-    ignore_index=None,
-):
+        input_ids,
+        image_thw_grids,
+        video_thw_grids,
+        target,
+        pad_token,
+        second_per_grid_ts,
+        ignore_index=None
+    ):
     """Build masks and position id for left to right model."""
     # Position ids. [3 X bs X seqlen]
     position_ids, _ = get_rope_index(
@@ -396,7 +335,7 @@ def get_ltor_masks_and_position_ids(
         image_grid_thw=image_thw_grids,
         video_grid_thw=video_thw_grids,
         second_per_grid_ts=second_per_grid_ts,
-        attention_mask=input_ids != pad_token,
+        attention_mask=input_ids != pad_token
     )
 
     # Loss mask.
@@ -410,7 +349,6 @@ def get_ltor_masks_and_position_ids(
 
     return attention_mask, loss_mask, position_ids
 
-
 def get_batch(data_iterator):
     """Generate a batch"""
     imgs = None
@@ -424,20 +362,12 @@ def get_batch(data_iterator):
     torch.cuda.nvtx.range_push("get_data")
     if data_iterator is not None and get_tensor_model_parallel_rank() == 0:
         data = next(data_iterator)
-        # pad_token_id = get_tokenizer().pad_token_id
-        pad_token_id = IGNORE_IDX
-        # while (data["target"] == pad_token_id).all() or (data["target"].shape[-1] < 986 or data["target"].shape[-1] > 1000): # for debug
-        while (data["target"] == pad_token_id).all():
-            logging.getLogger(__name__).warning(
-                "The current data is invalid because the target is all pad_token_id! Get next data to avoid fail, but it's better to check the data!"
-            )
-            data = next(data_iterator)
     else:
         data = None
 
-    data_text = broadcast_data(["text"], data, torch.int64)["text"]
+    data_text =  broadcast_data(["text"], data, torch.int64)["text"]
 
-    target = broadcast_data(["target"], data, torch.int64)["target"]
+    target =  broadcast_data(["target"], data, torch.int64)["target"]
     # shape: num_tiles x c x h x w
     imgs = broadcast_data(["imgs"], data, torch.float32)["imgs"]
 
@@ -446,23 +376,11 @@ def get_batch(data_iterator):
 
     # shape: n_image_samples
     image_thw_grids = broadcast_data(["image_thw_grids"], data, torch.long)["image_thw_grids"]
-
-    # global LAST_LARGE_IMG
-    # if LAST_LARGE_IMG:
-    #     torch.cuda.empty_cache()
-    #     LAST_LARGE_IMG=False
-    # if image_thw_grids.prod(axis=-1).sum() // 4 > 3000:
-    #     torch.cuda.empty_cache()
-    #     LAST_LARGE_IMG = True
-    args = get_args()
-    if data_text.shape[-1] == args.max_padding_length and get_pipeline_model_parallel_rank() == 0:
-        torch.cuda.empty_cache()
     # shape: n_video_samples
     video_thw_grids = broadcast_data(["video_thw_grids"], data, torch.long)["video_thw_grids"]
     # shape: n_video_samples
-    second_per_grid_ts = broadcast_data(['second_per_grid_ts'], data, torch.float32)[
-        'second_per_grid_ts'
-    ]
+    second_per_grid_ts = broadcast_data(['second_per_grid_ts'], data, torch.float32)['second_per_grid_ts']
+
 
     image_input_mask = broadcast_data(["image_input_mask"], data, torch.bool)["image_input_mask"]
     video_input_mask = broadcast_data(["video_input_mask"], data, torch.bool)["video_input_mask"]
@@ -480,7 +398,7 @@ def get_batch(data_iterator):
     # NOTE: no sequence packing in LLM inputs
     torch.cuda.nvtx.range_push("get_ltor_masks_and_position_ids")
     attention_mask, loss_mask, position_ids = get_ltor_masks_and_position_ids(
-        tokens, image_thw_grids, video_thw_grids, labels, IGNORE_IDX, second_per_grid_ts
+        tokens, image_thw_grids, video_thw_grids, labels, tokenizer.pad_token_id, second_per_grid_ts
     )
     torch.cuda.nvtx.range_pop()
 
@@ -495,75 +413,37 @@ def get_batch(data_iterator):
         image_thw_grids,
         video_thw_grids,
         image_input_mask,
-        video_input_mask,
+        video_input_mask
     )
 
-
-# define spiky loss as a loss that's 10x the max loss observed
-SPIKY_LOSS_FACTOR = 10
-
-
-def loss_func(
-    loss_mask: torch.Tensor, output_tensor: torch.Tensor, model: Optional[Qwen2_5VLModel] = None
-):
+def loss_func(loss_mask: torch.Tensor, output_tensor: torch.Tensor):
     """Loss function.
 
     Args:
         loss_mask (torch.Tensor): Used to mask out some portions of the loss
         output_tensor (torch.Tensor): The tensor with the losses
-        model (Qwen2_5VLModel, optional): The model (can be wrapped)
-
-    Returns:
-        the loss scalar for this micro-batch
-        the number of non-padded tokens in this microbatch
-        a dict containing reporting metrics on the loss and number of tokens across
-            the data parallel ranks
     """
     args = get_args()
 
-    if has_nvidia_modelopt and modelopt_args_enabled(args):  # [ModelOpt]
-        return loss_func_modelopt(loss_mask, output_tensor, model=model)
-
-    losses = output_tensor.view(-1).float()
+    losses = output_tensor.float()
     loss_mask = loss_mask.view(-1).float()
-    loss = torch.sum(losses * loss_mask)
+
+    loss = torch.stack([torch.sum(losses.view(-1) * loss_mask), loss_mask.sum()])
+    if args.context_parallel_size > 1:
+        torch.distributed.all_reduce(loss, group=mpu.get_context_parallel_group())
 
     # Check individual rank losses are not NaN prior to DP all-reduce.
-    rerun_state_machine = get_rerun_state_machine()
     if args.check_for_nan_in_loss_and_grad:
-        rerun_state_machine.validate_result(
-            result=loss,
-            rejection_func=torch.isnan,
-            message="found NaN in local forward loss calculation",
-            tolerance=0.0,  # forward pass calculations are determinisic
-            fatal=True,
-        )
-        rerun_state_machine.validate_result(
-            result=loss,
-            rejection_func=torch.isinf,
-            message="found Inf in local forward loss calculation",
-            tolerance=0.0,  # forward pass calculations are determinisic
-            fatal=True,
-        )
-    # Check for spiky loss
-    if args.check_for_spiky_loss:
-        rerun_state_machine.validate_result(
-            result=loss,
-            rejection_func=partial(
-                rerun_state_machine.is_unexpectedly_large,
-                threshold=SPIKY_LOSS_FACTOR,
-                context="loss",
-            ),
-            message="Spiky loss",
-            tolerance=0.0,  # forward pass calculations are determinisic
-            fatal=False,
+        global_rank = torch.distributed.get_rank()
+        assert not loss.isnan().any(), (
+            f"Rank {global_rank}: found NaN in local forward loss calculation. "
+            f"Device: {torch.cuda.current_device()}, node: {os.uname()[1]}"
         )
 
-    num_tokens = loss_mask.sum().clone().detach().to(torch.int)
-    reporting_loss = torch.cat([loss.clone().detach().view(1), num_tokens.view(1)])
-
-    return (loss, num_tokens, {'lm loss': reporting_loss})
+    averaged_loss = average_losses_across_data_parallel_group(loss)
+    averaged_loss = averaged_loss[0] / averaged_loss[1]
 
+    return loss[0] * args.context_parallel_size, {"lm loss": averaged_loss}
 
 def forward_step(data_iterator, model: Qwen2_5VLModel):
     """Forward training step.
@@ -572,51 +452,46 @@ def forward_step(data_iterator, model: Qwen2_5VLModel):
         data_iterator : Input data iterator
         model (GPTModel): The GPT Model
     """
-    args = get_args()
     timers = get_timers()
-
     # Get the batch.
-    timers('batch-generator', log_level=2).start()
-    global stimer
-    with stimer(bdata=True):
-        (
-            tokens,
-            labels,
-            loss_mask,
-            attention_mask,
-            position_ids,
-            imgs,
-            videos,
-            image_thw_grids,
-            video_thw_grids,
-            image_input_mask,
-            video_input_mask,
-        ) = get_batch(data_iterator)
-    timers('batch-generator').stop()
+    timers("batch-generator", log_level=2).start()
+    (
+        tokens,
+        labels,
+        loss_mask,
+        attention_mask,
+        position_ids,
+        imgs,
+        videos,
+        image_thw_grids,
+        video_thw_grids,
+        image_input_mask,
+        video_input_mask
+    ) = get_batch(data_iterator)
+    timers("batch-generator").stop()
+
     vision_data = torch.cat([imgs, videos], dim=0)
     vision_grid = torch.cat([image_thw_grids, video_thw_grids], dim=0)
-    with stimer:
-        output_tensor = model(
-            input_ids=tokens,
-            position_ids=position_ids,
-            vision_data=vision_data,
-            vision_grid_thw=vision_grid,
-            video_start_index=image_input_mask.sum().cpu().item(),
-            image_input_mask=image_input_mask,
-            video_input_mask=video_input_mask,
-            attention_mask=attention_mask,
-            labels=labels,
-        )
 
-    return output_tensor, partial(loss_func, loss_mask, model=model)
+    output_tensor = model(
+        input_ids = tokens,
+        position_ids = position_ids,
+        vision_data = vision_data,
+        vision_grid_thw =  vision_grid,
+        video_start_index = image_input_mask.sum().cpu().item(),
+        image_input_mask = image_input_mask,
+        video_input_mask = video_input_mask,
+        attention_mask = attention_mask,
+        labels = labels
+    )
 
+    return output_tensor, partial(loss_func, loss_mask)
 
 def run_online_eval(model):
     """Run an evaluation benchmark during training."""
     # Do nothing.
     return []
 
-
 def write_online_eval_to_tensorboard(data, iteration, writer):
     """Write online evaluation data to Tensorboard."""
     if not writer:
@@ -626,7 +501,6 @@ def write_online_eval_to_tensorboard(data, iteration, writer):
         for k, v in item.items():
             writer.add_scalar(k, v, iteration)
 
-
 def datasets_provider(worker_config=None):
     """Create multimodal train, validation and test datasets."""
     args = get_args()
@@ -636,76 +510,70 @@ def datasets_provider(worker_config=None):
         batch_size=args.micro_batch_size,
         task_encoder=TaskEncoder(),
         worker_config=worker_config,
-        virtual_epoch_length=0,
-        max_samples_per_sequence=args.max_samples_per_sequence,  # sequential shuffle in a tar
-        shuffle_buffer_size=args.shuffle_buffer_size,  # shuffle in a sequential
+        virtual_epoch_length=1000,
+        max_samples_per_sequence=100,
+        shuffle_buffer_size=100,
+        handler=print_error_handler,
+        image_decode="pil",
+    )
+
+    val_datasets = get_val_datasets(
+        dname,
+        batch_size=args.micro_batch_size,
+        # This is the total number over all workers
+        # limit=args.eval_iters * get_num_microbatches(),
+        task_encoder=TaskEncoder(),
+        worker_config=worker_config,
         handler=print_error_handler,
-        repeat=True,
         image_decode="pil",
     )
-    val_datasets_without_source_datasets = None
-    if args.eval_iters > 0:
-        val_datasets = get_val_datasets(
-            dname,
-            batch_size=args.micro_batch_size,
-            # This is the total number over all workers
-            # limit=args.eval_iters * get_num_microbatches(),
-            task_encoder=TaskEncoder(),
+    val_datasets_without_source_datasets = [
+        # Limit the dataset to eval_iters * num_microbatches
+        LimitDataset(
+            # Repeat the inner dataset in case it's too short
+            RepeatDataset(val_ds, worker_config=worker_config),
+            length=args.eval_iters * get_num_microbatches(),
             worker_config=worker_config,
-            handler=print_error_handler,
-            image_decode="pil",
+            reset_after_epoch=True,
         )
-        val_datasets_without_source_datasets = [
-            # Limit the dataset to eval_iters * num_microbatches
-            LimitDataset(
-                # Repeat the inner dataset in case it's too short
-                RepeatDataset(val_ds, worker_config=worker_config),
-                length=args.eval_iters * get_num_microbatches(),
-                worker_config=worker_config,
-                reset_after_epoch=True,
-            )
-            for val_ds, _src_ds in val_datasets
-        ]
+        for val_ds, _src_ds in val_datasets
+    ]
 
     return train_dataset, val_datasets_without_source_datasets, None
 
-
-def is_first_or_last_stage(pp_size, transformer_pipeline_model_parallel_size):
+def is_first_or_last_stage(pp_size, encoder_pipeline_model_parallel_size):
     """Check if the current pipeline parallel stage is the first or last stage."""
-    if pp_size == 1:  # No pipeline parallelism.
+    if pp_size == 1:    # No pipeline parallelism.
         return True
 
     is_valid_rank = False
     pp_rank = get_pipeline_model_parallel_rank()
-    if transformer_pipeline_model_parallel_size == 0:
+    if encoder_pipeline_model_parallel_size == 0:
         # No separate pipeline stage for the vision model. Run the dataloader on the first and last pipeline stage.
-        is_valid_rank = pp_rank in (0, pp_size - 1)
-    elif transformer_pipeline_model_parallel_size == 1:
+        is_valid_rank = pp_rank in (0, pp_size-1)
+    elif encoder_pipeline_model_parallel_size == 1:
         # Separate pipeline stage for the vision model. Run the dataloader on the first vision and LM stage and last LM stage.
-        is_valid_rank = pp_rank in (0, 1, pp_size - 1)
+        is_valid_rank = pp_rank in (0, 1, pp_size-1)
     else:
         raise NotImplementedError("encoder-pipeline-model-parallel-size > 1 is not supported yet")
 
     return is_valid_rank
 
-
-def is_dataloader_rank(transformer_pipeline_model_parallel_size):
+def is_dataloader_rank(encoder_pipeline_model_parallel_size):
     """Check if we should have the dataloader on this tensor and pipeline parallel rank."""
     # Run dataloader only on the first tensor parallel rank (will be broadcasted to others).
     is_first_rank = get_tensor_model_parallel_rank() == 0
 
-    # NOTE(lizhiyu): when pp_size > 2
-    # pp_size = get_pipeline_model_parallel_world_size()
-    # is_first_rank = is_first_rank and is_first_or_last_stage(pp_size, transformer_pipeline_model_parallel_size)
+    pp_size = get_pipeline_model_parallel_world_size()
+    is_first_rank = is_first_rank and is_first_or_last_stage(pp_size, encoder_pipeline_model_parallel_size)
 
     return is_first_rank
 
-
 def train_valid_test_dataloaders_provider(train_val_test_num_samples):
     """Build multimodal train, validation and test dataloaders."""
     args = get_args()
     # Dataloader is only on specific ranks.
-    if not is_dataloader_rank(args.transformer_pipeline_model_parallel_size):
+    if not is_dataloader_rank(args.encoder_pipeline_model_parallel_size):
         return None, None, None
 
     worker_debug_path = None
@@ -732,34 +600,27 @@ def train_valid_test_dataloaders_provider(train_val_test_num_samples):
             data_save_name = get_checkpoint_name(
                 args.dataloader_save,
                 args.iteration,
-                pipeline_rank=0,  # Only the first pipeline parallel rank stores the dataloader checkpoint.
+                pipeline_rank=0,    # Only the first pipeline parallel rank stores the dataloader checkpoint.
                 basename=f"train_dataloader_dprank{dp_rank:03d}.pt",
             )
             if os.path.exists(data_save_name):
                 try:
-                    dataset_state_dict = torch.load(
-                        data_save_name, map_location="cpu", weights_only=False
-                    )
+                    dataset_state_dict = torch.load(data_save_name, map_location="cpu", weights_only=False)
                     train_dataloader.restore_state_rank(dataset_state_dict["dataloader_state_dict"])
                     print_rank_0(f"restored dataset state from {data_save_name}")
                 except Exception as e:
                     print_rank_0("loading dataloader checkpoint failed. Skipping. " + str(e))
 
-    if valid_ds1 is not None:
-        valid_dataloader = [
-            EnergonDataloader(get_loader(valid_ds, worker_config=worker_config))
-            for valid_ds in valid_ds1
-        ]
-    else:
-        valid_dataloader = EnergonDataloader(None)
-    test_dataloader = None  # NOTE: no test
+    valid_dataloader = [
+        EnergonDataloader(get_loader(valid_ds, worker_config=worker_config))
+        for valid_ds in valid_ds1
+    ]
+    test_dataloader = None # NOTE: no test
 
     return EnergonDataloader(train_dataloader), valid_dataloader, EnergonDataloader(test_dataloader)
 
-
 class EnergonDataloader:
     """A wrapper to use Megatron Energon dataloader with the Megatron-LM training loop."""
-
     def __init__(self, dataloader):
         self._dataloader = dataloader
         self._iter = iter(cyclic_iter(dataloader))
@@ -783,12 +644,7 @@ def cyclic_iter(iter):
 def add_multimodal_extra_args(parser):
     """Extra arguments."""
     group = parser.add_argument_group(title="multimodal arguments")
-    group.add_argument(
-        "--disable-vision-class-token",
-        action="store_true",
-        default=False,
-        help="Disable vision class token",
-    )
+    group.add_argument("--disable-vision-class-token", action="store_true", default=False, help="Disable vision class token")
     group.add_argument(
         "--dataloader-save", type=str, default=None, help="Energon dataloader state save path"
     )
@@ -799,49 +655,6 @@ def add_multimodal_extra_args(parser):
     group.add_argument("--temporal-patch-size", type=int, default=2)
     group.add_argument("--patch-size", type=int, default=14)
     group.add_argument("--max-padding-length", type=int, default=2048)
-    group.add_argument(
-        "--enable-variable-seq-lengths",
-        action="store_true",
-        default=False,
-        help="Enable variable sequence lengths",
-    )
-    group.add_argument(
-        "--vision-root", type=str, default=None, help="The vision dirctory root path."
-    )
-    group.add_argument(
-        "--max-samples-per-sequence",
-        type=int,
-        default=2**31 - 1,
-        help="Max sequencial seqence samples in a slice",
-    )
-    group.add_argument(
-        "--shuffle-buffer-size",
-        type=int,
-        default=0,
-        help="The buffer size to shuffle the samples in a seqence",
-    )
-    # learning rate
-    group.add_argument(
-        "--vision-ration",
-        type=float,
-        default=0.1,
-        help="The learning rate ration of vision(inlude merger) compared with llm",
-    )
-    group.add_argument(
-        "--image-max-pixels",
-        type=int,
-        default=768 * 768,
-        help="The maximum pixels of a single image",
-    )
-    group.add_argument(
-        "--image-min-pixels", type=int, default=32 * 32, help="The minimum pixels of a single image"
-    )
-    group.add_argument(
-        "--vision-recompute-layer-steps",
-        type=int,
-        default=0,
-        help="The recmoute layers for vision using uniform method. 0 is disable.",
-    )
 
     # just for checkpoint conversion
     group.add_argument(
@@ -852,21 +665,15 @@ def add_multimodal_extra_args(parser):
             "If False, convert a Transformers checkpoint to a Megatron checkpoint."
         ),
     )
-    group.add_argument(
-        "--freeze-LM", action="store_true", default=False, help="Freeze the language model"
-    )
-    group.add_argument(
-        "--freeze-ViT", action="store_true", default=False, help="Freeze the vision model"
-    )
+    group.add_argument("--freeze-LM", action="store_true", default=False, help="Freeze the language model")
+    group.add_argument("--freeze-ViT", action="store_true", default=False, help="Freeze the vision model")
     group.add_argument(
         "--allow-missing-vision-projection-checkpoint",
         action="store_true",
         default=False,
         help="Allow missing vision projection checkpoint",
     )
-    group.add_argument(
-        "--use-te", action="store_true", default=False, help="Use transformer engine"
-    )
+    group.add_argument("--use-te", action="store_true", default=False, help="Use transformer engine")
     return parser
 
 

