diff --git a/flagscale/train/models/qwen3_vl/language_transformer_block.py b/flagscale/train/models/qwen3_vl/language_transformer_block.py
old mode 100755
new mode 100644
index 0a86705c..7b6e61da
--- a/flagscale/train/models/qwen3_vl/language_transformer_block.py
+++ b/flagscale/train/models/qwen3_vl/language_transformer_block.py
@@ -13,7 +13,7 @@ from megatron.core.transformer.transformer_block import TransformerBlock
 from megatron.core.fusions.fused_layer_norm import FusedLayerNorm
 from megatron.core import parallel_state, tensor_parallel
 from megatron.core.enums import Fp8Recipe
-from megatron.core.fp4_utils import get_fp4_context
+# from megatron.core.fp4_utils import get_fp4_context
 from megatron.core.fp8_utils import get_fp8_context
 from megatron.core.inference.contexts import BaseInferenceContext
 from megatron.core.packed_seq_params import PackedSeqParams
@@ -286,10 +286,10 @@ class LanguageTransformerBlock(TransformerBlock):
             outer_quantization_context = (
                 get_fp8_context(self.config) if use_outer_quantization_context else nullcontext()
             )
-        elif self.config.fp4:
-            use_outer_quantization_context = False
-            use_inner_quantization_context = True
-            outer_quantization_context = nullcontext()
+        # elif self.config.fp4:
+        #     use_outer_quantization_context = False
+        #     use_inner_quantization_context = True
+        #     outer_quantization_context = nullcontext()
         else:
             # No quantization
             use_outer_quantization_context = False

