diff --git a/dcu_megatron/training/arguments.py b/dcu_megatron/training/arguments.py
new file mode 100644
index 000000000..ffb158d79
--- /dev/null
+++ b/dcu_megatron/training/arguments.py
@@ -0,0 +1,148 @@
+import os
+import argparse
+
+from typing import Union
+from megatron.training.arguments import add_megatron_arguments
+from megatron.core.msc_utils import MultiStorageClientFeature
+
+
+def remove_original_params(parser, param_names: Union[list, str]):
+    if isinstance(param_names, str):
+        param_names = [param_names]
+
+    for action in parser._actions:
+        if action.dest in param_names:
+            parser._actions.remove(action)
+            for option_string in action.option_strings:
+                if option_string in parser._option_string_actions:
+                    del parser._option_string_actions[option_string]
+
+
+def add_megatron_arguments_patch(parser: argparse.ArgumentParser):
+    parser = add_megatron_arguments(parser)
+
+    # add extra arguments
+    parser = _add_extra_network_size_args(parser)
+    parser = _add_extra_training_args(parser)
+    parser = _add_extra_initialization_args(parser)
+    parser = _add_extra_distributed_args(parser)
+    # parser = _add_extra_tokenizer_args(parser)
+
+    return parser
+
+
+def parse_args(extra_args_provider=None, ignore_unknown_args=False):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser(description='Megatron-LM Arguments',
+                                     allow_abbrev=False)
+
+    parser = add_megatron_arguments_patch(parser)
+
+    # Custom arguments.
+    if extra_args_provider is not None:
+        parser = extra_args_provider(parser)
+
+    # Parse.
+    if ignore_unknown_args:
+        args, _ = parser.parse_known_args()
+    else:
+        args = parser.parse_args()
+
+    # Experimental yaml
+    if args.yaml_cfg is not None:
+        from megatron.training.yaml_arguments import load_yaml
+        assert args.yaml_cfg and not args.use_legacy_models, \
+            "Yaml config is not supported with legacy models."
+        args = load_yaml(args.yaml_cfg)
+
+    # Args from environment
+    # args.rank = int(os.getenv('RANK', '0'))
+    # args.world_size = int(os.getenv("WORLD_SIZE", '1'))
+
+    # Args to disable MSC
+    if not args.enable_msc:
+        MultiStorageClientFeature.disable()
+        assert MultiStorageClientFeature.is_enabled() is False
+        print('WARNING: The MSC feature is disabled.')
+
+    # Args from environment
+    args.rank = int(os.getenv('RANK', '0'))
+    args.world_size = int(os.getenv("WORLD_SIZE", '1'))
+
+    return args
+
+
+def _add_extra_network_size_args(parser):
+    # 删除原参数
+    remove_original_params(parser, ["normalization"])
+
+    # 重定义参数
+    group = parser.add_argument_group(title='extra network size args')
+    group.add_argument('--normalization', default='LayerNorm',
+                       choices=['LayerNorm', 'RMSNorm', 'LightopRMSNorm'],
+                       help='Which normalization technique to use.')
+    return parser
+
+
+def _add_extra_distributed_args(parser):
+    group = parser.add_argument_group(title='extra distributed args')
+    group.add_argument('--rank', default=-1, type=int,
+                       help='node rank for distributed training')
+    group.add_argument('--world-size', type=int, default=8,
+                       help='number of nodes for distributed training')
+    group.add_argument('--dist-url',
+                       help='Which master node url for distributed training.')
+    return parser
+
+
+def _add_extra_training_args(parser):
+    group = parser.add_argument_group(title='extra training args')
+    group.add_argument('--use-hip-profiler', action='store_true',
+                       help='Use HIP PROFILER',
+                       dest='use_hip_profiler')
+    group.add_argument('--profile-dir', type=str, default="./",
+                       help='profile dir to save.')
+
+    return parser
+
+
+def _add_extra_initialization_args(parser):
+    group = parser.add_argument_group(title='extra initialization args')
+    group.add_argument('--reproduce', action='store_true',
+                       help='reproduce train loss, need set --seed > 0.')
+
+    return parser
+
+
+def _add_extra_tokenizer_args(parser):
+    # 删除原参数
+    remove_original_params(parser, ["tokenizer_type"])
+
+    # 重定义参数
+    group = parser.add_argument_group(title='extra tokenizer args')
+    group.add_argument('--extra-vocab-size', type=int, default=0,
+                       help="--extra-vocab-size")
+    group.add_argument('--tokenizer-type', type=str,
+                       default=None,
+                       choices=['BertWordPieceLowerCase',
+                                'BertWordPieceCase',
+                                'GPT2BPETokenizer',
+                                'SentencePieceTokenizer',
+                                'GPTSentencePieceTokenizer',
+                                'HuggingFaceTokenizer',
+                                'Llama2Tokenizer',
+                                'Llama3Tokenizer',
+                                'QwenTokenizer',
+                                'TikTokenizer',
+                                'MultimodalTokenizer',
+                                'NullTokenizer',
+                                'DeepSeekV2Tokenizer'],
+                       help='What type of tokenizer to use.')
+    return parser
+
+
+def _add_flux_args(parser):
+    group = parser.add_argument_group(title='flux args')
+    group.add_argument('--flux-transpose-weight', action='store_true', default=False,
+                       help='Whether to transpose weight when using flux kernel')
+    return parser

